File: apps/frontend/lib/api/hostel.ts
import { APIResponse, Hostel } from "../dto";
import { CreateHostelListingData } from "../forms";
import { UploadFile } from "../utils";
import {
  AreaOptions,
  areaSerializer,
  HostelFilterOptions,
  hostelFilterSerializer,
} from "./filter";
import { PaginationOptions, paginationSerializer } from "./pagination";
import { uploadPhoto } from "./upload";

export async function createHostelListing(
  data: CreateHostelListingData,
  photoUrls: UploadFile[] | null,
) {
  if (!photoUrls || photoUrls.length === 0) {
    throw new Error("Photo is required");
  }

  const photoPromises = photoUrls.map(photo =>
    uploadPhoto({
      entityName: data.name,
      entityType: "hostel",
      primaryPhoto: photo.file,
    }),
  );

  const publicUrls = await Promise.all(photoPromises);

  const dataWithPhotosAndAmenities = {
    ...data,
    photo_urls: publicUrls,
    is_verified: data.isVerified,
    amenities: data.amenities.map(amenity => amenity.text),
  };

  const res = await fetch(`${process.env.NEXT_PUBLIC_API_URL}/api/v1/hostels`, {
    method: "POST",
    headers: {
      "Content-Type": "application/json",
    },
    body: JSON.stringify(dataWithPhotosAndAmenities),
  });

  const response = await res.json();
  if (!response.success) {
    throw new Error(response.message);
  }

  return response as any as APIResponse<CreateHostelListingData>;
}

export async function fetchHostel(slug: string): Promise<APIResponse<Hostel> | null> {
  try {
    const response = await fetch(
      `${process.env.NEXT_PUBLIC_API_URL}/api/v1/hostels/${slug}`,
    );

    if (!response.ok) {
      if (response.status === 404) return null;
      throw new Error("Failed to fetch hostel");
    }

    return response.json() as any as APIResponse<Hostel>;
  } catch (error) {
    console.error(error);
    return null;
  }
}

export async function fetchHostelsByArea({
  areaFilters,
  hostelFilters,
  paginationFilters,
}: {
  areaFilters: AreaOptions;
  hostelFilters: HostelFilterOptions;
  paginationFilters: PaginationOptions;
}): Promise<APIResponse<Hostel[]>> {
  try {
    const paginationParams = paginationSerializer({
      page: paginationFilters.page,
      limit: paginationFilters.limit,
    });

    const areaParams = areaSerializer({
      areaType: areaFilters.areaType,
      areaId: areaFilters.areaId,
    });

    const hostelFilterParams = hostelFilterSerializer({
      sortBy: hostelFilters.sortBy,
      minPrice: hostelFilters.minPrice,
      maxPrice: hostelFilters.maxPrice,
      isVerified: hostelFilters.isVerified,
    });
    const response = await fetch(
      `${process.env.NEXT_PUBLIC_API_URL}/api/v1/hostels/search${areaParams}${hostelFilterParams && "&" + hostelFilterParams.replace("?", "")}${paginationParams && "&" + paginationParams.replace("?", "")}`,
    );

    if (!response.ok) {
      throw new Error("Failed to fetch hostels by area");
    }

    return response.json() as any as APIResponse<Hostel[]>;
  } catch (error) {
    throw new Error("Failed to fetch hostels by area");
  }
}

File: apps/backend/internal/dto/hostel.go
package dto

type CreateHostel struct {
	Name                string   `json:"name" validate:"required,min=2,max=100"`
	Address             string   `json:"address" validate:"required,min=2,max=255"`
	Description         string   `json:"description" validate:"required,min=10,max=4000"`
	Neighborhood        string   `json:"neighborhood" validate:"required,min=2,max=100"`
	NeighborhoodID      string   `json:"neighborhoodId" validate:"required,uuid"`
	EstimatedPriceRange float64  `json:"estimatedPriceRange" validate:"required"`
	Latitude            float64  `json:"latitude" validate:"required,min=-90,max=90"`
	Longitude           float64  `json:"longitude" validate:"required,min=-180,max=180"`
	PhotoURLs           []string `json:"photo_urls" validate:"required"`
	IsVerified          bool     `json:"is_verified" validate:"required"`
	Amenities           []string `json:"amenities" validate:"required"`
}

type Hostel struct {
	Name                string   `json:"name"`
	Address             string   `json:"address"`
	Description         string   `json:"description"`
	Neighborhood        string   `json:"neighborhood"`
	EstimatedPriceRange float64  `json:"estimatedPriceRange"`
	Latitude            float64  `json:"latitude"`
	Longitude           float64  `json:"longitude"`
	PhotoURLs           string   `json:"photo_urls"`
	Slug                string   `json:"slug"`
	IsVerified          bool     `json:"isVerified"`
	Amenities           []string `json:"amenities"`
}
